drivers: crypto:  Support for hashing

ceolin opened the pull request:
- Cleanup and organize crypto drivers 
  - Proper namespaces (crypto and cipher used properly)
  - Rename files
- Add API for hashing
- Add support for hashing (sha-2) on mbedTLS shim driver
- Add a new test for hashing

Some items are similar to cipher but others are not. To keep it simple and to not make is more complicated a new set of API and data structure were added. Re-use some of the APIs would result and call with not used parameters which is not good.
The goal with this pr is to support hash capabilities offered by some HW. Currently there is no way to expose it. For long term, we should look psa_crypto API  as a possible replacement for it.

Commit:
drivers: crypto: Add cipher namespace in some API

Add 'cipher' namespace in some in the driver API since these
operations are for cipher.

Set a namespace to make it clear that these are cipher operations,
this allow further functionalities, like hash, to be added in this
driver API.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
drivers: crypto: Change HW capabilities query namespace

This is not cipher specific, just replace it crypto to make it clear.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
drivers: crypto: Organize documentation

Move the Cipher subgroup documentation to where this API is set.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
drivers: crypto: Rename cipher file

This file defines the crypto driver API, cipher is supposed to be just
one type of capability (other can be hash) of these drivers, just
change the file name to be consistent with it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
drivers: crypto: s/cipher_struct/cipher

Just give a better name to this file since now we have changed the
file where crypto driver API is defined.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
drivers: crypto: Add hash support

Several HW with crypto accelerators have also hash capabilities.
The semantic for hashing is different of ciphers, so this commit
introduce support for hashing within crypto drivers.

The API is following the driver with the operation assigned /
associated with the session.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
drivers: crypto: mtls: Add support for hashing

Add support for hashing on mtls driver.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
modules: mbedTLS: Add Kconfig options for hashing

Add explicit Kconfig options to enable SHA support on mbedTLS.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
drivers: crypto: Keep namespace consistent

The API to set a callback has the namespace cipher but the driver
function pointer was using the namespace crypto. As this API belongs
to the cipher subgroup, just rename the function pointer in the driver
to be consistent.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: crypto: Add test for hashing on crypto driver

Test crypto driver hash capability.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
drivers: crypto: Move capabilities to crypto.h

Capabilities are for the driver, though most of options apply only
for ciphers some of them also apply for hashes.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

d3zd3z reviewed (commented):
I think this is ok, but hash is going to be a lot more useful if we provide a way to incrementally update the hash state. A lot of uses of hash functions will need this, and without it, would have to allocate large buffers to store the intermediate result.

Commit:
drivers: crypto: Add multipart hash support

Add a new API to support multipart hash calculation. The API allows
split the data input to be split in small chunks.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
> I think this is ok, but hash is going to be a lot more useful if we provide a way to incrementally update the hash state. A lot of uses of hash functions will need this, and without it, would have to allocate large buffers to store the intermediate result.

indeed. Just add an API (similar to mbedTLS and PSA crypto) that updates the input data.

teburd reviewed (approved):
This looks like a really nice changeset to me.

One thing I did note, was the lack of an output len member in the hash_pkt struct. The hash output size should be known by any author using this when writing their code, of course! 

That said, it wouldn't cost much to demand the output buffer size be a hash_pkt member and assert on it when possible to avoid unwanted surprises.
